### PR TITLE
Using MongoDB for storing Tiles and Metric data

### PR DESCRIPTION
**Pull request by jjankowski87**

### CODE REVIEW

File: MetricsDashboard.Contract/BooleanResultDto.cs [hunk 1]
```diff
@@ -0,0 +1,11 @@
+﻿using System;
+
+namespace MetricsDashboard.Contract
+{
+    public class BooleanResultDto
```
Comment (Carq) [inline MetricsDashboard.Contract/BooleanResultDto.cs]: We need to think about better name of this kind of tile, something lige StatusTile maybe?

File: MetricsDashboard.Core/ITileRepository.cs [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+using System.Threading;
+using System.Threading.Tasks;
+using MetricsDashboard.Core.Entities;
+using MetricsDashboard.Core.Models;
+
+namespace MetricsDashboard.Core
+{
+    public interface ITileRepository
+    {
+        Task<TileEntity> GetTileAsync(string tileName, CancellationToken cancellationToken);
```
Comment (Carq) [inline MetricsDashboard.Core/ITileRepository.cs:10]: If we want to use Domain Models then repositories should return DomainModels, and MongoEntities should be used only inside of repositories.

File: MetricsDashboard.Core/Models/BooleanTile.cs [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+using MetricsDashboard.Contract.Enums;
+using MetricsDashboard.Core.Entities;
+
+namespace MetricsDashboard.Core.Models
+{
+    public class BooleanTile : ITile
+    {
+        private readonly string _name;
+
+        private readonly MetricEntity<bool> _latest;
```
Comment (Carq) [inline MetricsDashboard.Core/Models/BooleanTile.cs]: In this approach Mongo Entities shouldnt be part of Domain Model.

File: MetricsDashboard.Core/Models/BooleanTile.cs [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+
+        public BooleanTile(string name, IEnumerable<MetricEntity<bool>> metrics)
+        {
+            _name = name;
+            _latest = metrics.OrderByDescending(m => m.AddedOn).First();
+        }
+
+        public TileDto ToDto()
```
Comment (Carq) [inline MetricsDashboard.Core/Models/BooleanTile.cs]: I would recommended to use some kind of mapper or ConverterClass to do make convention between types.

File: MetricsDashboard.Core/Models/IDatabaseSettings.cs [hunk 5]
```diff
@@ -0,0 +1,9 @@
+﻿namespace MetricsDashboard.Core.Models
```
Comment (Carq) [inline MetricsDashboard.Core/Models/IDatabaseSettings.cs]: Models -> DomainModels

File: MetricsDashboard.Core/Models/IDatabaseSettings.cs [hunk 6]
```diff
@@ -0,0 +1,9 @@
...
+{
+    public interface IDatabaseSettings
```
Comment (Carq) [inline MetricsDashboard.Core/Models/IDatabaseSettings.cs]: This should be in "Configuration" folder

File: MetricsDashboard.Core/TileDataService.cs [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+            await SaveTileDataAsync(saveValueDto, TileType.Boolean, cancellationToken);
+        }
+
+        private async Task SaveTileDataAsync<TValue>(
+            SaveValueDto<TValue> saveValueDto,
+            TileType tileType,
+            CancellationToken cancellationToken)
+        {
+            var tile = await _tileRepository.GetTileAsync(saveValueDto.Name, cancellationToken);
+            if (tile == null)
```
Comment (Carq) [inline MetricsDashboard.Core/TileDataService.cs]: We introduce this kind of class
```csharp
   public static class ObjectExtensions
    {
        public static bool Exists(this object singleObject)
        {
            return singleObject != null;
        }

        public static bool NotExists(this object singleObject)
        {
            return !singleObject.Exists();
        }
    }
```
And thanks to that we will have more readable code:
```csharp
if (tile.NotExists())
{
// code...
}

**Review by Carq — CHANGES_REQUESTED** [review]

File: MetricsDashboard.Core/ITileRepository.cs (hunk 2 above)
Comment (Carq) [reply MetricsDashboard.Core/ITileRepository.cs:10]: Thanks to that we will separate Domain Code from Data Access Code.

File: MetricsDashboard.Core/Entities/TileEntity.cs [hunk 8]
```diff
@@ -0,0 +1,16 @@
+﻿using MetricsDashboard.Contract.Enums;
+using MongoDB.Bson;
+using MongoDB.Bson.Serialization.Attributes;
+
+namespace MetricsDashboard.Core.Entities
```
Comment (Carq) [inline MetricsDashboard.Core/Entities/TileEntity.cs:5]: Entities -> MongoEntities?

File: MetricsDashboard.Contract/BooleanResultDto.cs (hunk 1 above)
Comment (jjankowski87) [reply MetricsDashboard.Contract/BooleanResultDto.cs]: renamed to StatusDto

File: MetricsDashboard.Core/ITileRepository.cs (hunk 2 above)
Comment (jjankowski87) [reply MetricsDashboard.Core/ITileRepository.cs:10]: Now repository returns only Entities

File: MetricsDashboard.Core/Models/BooleanTile.cs (hunk 3 above)
Comment (jjankowski87) [reply MetricsDashboard.Core/Models/BooleanTile.cs]: removed Domain Models

File: MetricsDashboard.Core/Models/BooleanTile.cs (hunk 4 above)
Comment (jjankowski87) [reply MetricsDashboard.Core/Models/BooleanTile.cs]: created TileMapper class

File: MetricsDashboard.Core/Models/IDatabaseSettings.cs (hunk 5 above)
Comment (jjankowski87) [reply MetricsDashboard.Core/Models/IDatabaseSettings.cs]: removed

File: MetricsDashboard.Core/Models/IDatabaseSettings.cs (hunk 6 above)
Comment (jjankowski87) [reply MetricsDashboard.Core/Models/IDatabaseSettings.cs]: moved to Core.Configuration

File: MetricsDashboard.Core/TileDataService.cs (hunk 7 above)
Comment (jjankowski87) [reply MetricsDashboard.Core/TileDataService.cs]: cool, added

File: MetricsDashboard.Core/Entities/TileEntity.cs (hunk 8 above)
Comment (jjankowski87) [reply MetricsDashboard.Core/Entities/TileEntity.cs:5]: Are there any other kind of entities? I think just entities would be sufficient.

**Review by Carq — APPROVED** [review]

**Pull request merged**